### PR TITLE
Clean packages from primary.xml

### PR DESCRIPTION
**Pull request by olenagerasimova**
For #138 created `XmlPrimaryMaid` that can clean packages from primary.xml without keeping whole xml in memory.

### CODE REVIEW

**Comment by g4s8** [comment]
@0crat status

**Comment by 0crat** [comment]
> @0crat status ([here](https://github.com/artipie/rpm-adapter/issues/139#issuecomment-618380360))

@g4s8 This is what I know about this job in CT2E6TK9B, as in [§32](http://www.zerocracy.com/policy.html#32):

  * The job [#139](https://github.com/artipie/rpm-adapter/issues/139) is [in scope](http://www.zerocracy.com/policy.html#14) for 1hr
  * The role is `REV`
  * The job is not assigned to anyone
  * The [budget](http://www.zerocracy.com/policy.html#4) is not set yet
  * Job [footprint](https://www.0crat.com/footprint/CT2E6TK9B?q=%7Bjob%3A%27gh%3Aartipie%2Frpm-adapter%23139%27%2Ctype%3A%7B%24not%3A%2F%5ENotify%2F%7D%7D) (restricted area)

**Comment by 0crat** [comment]
This pull request [#139](https://github.com/artipie/rpm-adapter/issues/139) is assigned to @vatavuk[/z](https://www.0crat.com/u/vatavuk), here is [why](https://www.0crat.com/footprint/CT2E6TK9B/5611f965-4f8f-409d-b7bc-c20c660565f4); the budget is 15 minutes, see [§4](http://www.zerocracy.com/policy.html#4); please, read [§27](http://www.zerocracy.com/policy.html#27) and when you decide to accept the changes, inform @g4s8[/z](https://www.0crat.com/u/g4s8) (the architect) right in this ticket; if you decide that this PR should not be accepted ever, also inform the architect; this [blog post](http://www.yegor256.com/2015/02/09/serious-code-reviewer.html) will help you understand what is expected from a code reviewer; there will be a monetary reward for this job

**Comment by Vatavuk** [comment]
@olenagerasimova refuse

**Comment by olenagerasimova** [comment]
@Vatavuk are you refusing PR (like it should never be merged) or to review it?

**Comment by g4s8** [comment]
@0crat assign me

**Comment by 0crat** [comment]
@0crat[/z](https://www.0crat.com/u/0crat) I see that you unassigned this issue; the order is still assigned to @g4s8[/z](https://www.0crat.com/u/g4s8) though; to cancel the order use `refuse`, as in [§6](http://www.zerocracy.com/policy.html#6)

**Comment by 0crat** [comment]
> @0crat assign me ([here](https://github.com/artipie/rpm-adapter/issues/139#issuecomment-618846014))

@g4s8 This pull request [#139](https://github.com/artipie/rpm-adapter/issues/139) is assigned to @g4s8[/z](https://www.0crat.com/u/g4s8), here is [why](https://www.0crat.com/footprint/CT2E6TK9B/5df51069-e5fc-4daf-84f6-e47b62016e44); the budget is 15 minutes, see [§4](http://www.zerocracy.com/policy.html#4); please, read [§27](http://www.zerocracy.com/policy.html#27) and when you decide to accept the changes, inform @g4s8[/z](https://www.0crat.com/u/g4s8) (the architect) right in this ticket; if you decide that this PR should not be accepted ever, also inform the architect; this [blog post](http://www.yegor256.com/2015/02/09/serious-code-reviewer.html) will help you understand what is expected from a code reviewer; there will be no monetary reward for this job

File: src/main/java/com/artipie/rpm/pkg/PrimaryOutput.java [hunk 1]
```diff
@@ -33,9 +33,8 @@
 /**
  * Package output for {@code primary} metadata.
  * @since 0.4
- * @todo #124:30min Modify this class and XmlPrimary to be able to work with already existing
- *  primary.xml: we need to accept this this xml data about new packages and delete data about
- *  package by package id (checksum). Do fot forget about tests.
+ * @todo #138:30min Make this class be able to work this already existing primary by
```
Comment (g4s8) [inline src/main/java/com/artipie/rpm/pkg/PrimaryOutput.java:36]: @olenagerasimova don't you need to mention how to use new `XmlPrimaryMaid` in `PrimaryOutput`?

File: src/main/java/com/artipie/rpm/meta/XmlPrimaryMaid.java [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+        this.source = source;
+        this.target = target;
+    }
+
+    /**
+     * Clears records about packages by given checksums (checksums).
+     * @param checksums What to clear
+     * @throws IOException When smth wrong
+     */
+    void clean(final List<String> checksums) throws IOException {
```
Comment (g4s8) [inline src/main/java/com/artipie/rpm/meta/XmlPrimaryMaid.java:76]: @olenagerasimova this class is public, but method is package private, I think it should have the same visibility

File: src/main/java/com/artipie/rpm/meta/XmlPrimaryMaid.java [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+     * Where to write clean data.
+     */
+    private final Path target;
+
+    /**
+     * Ctor.
+     * @param source What to clear
+     * @param target Where to write
+     */
+    public XmlPrimaryMaid(final Path source, final Path target) {
```
Comment (g4s8) [inline src/main/java/com/artipie/rpm/meta/XmlPrimaryMaid.java:66]: @olenagerasimova does client code need to specify both `source` and `target`? If the use case of this class is to clean packages from `source`, I'd remove `target` field and create temporary file right in `clean` method and replace `source` with this file on success

File: src/main/java/com/artipie/rpm/pkg/PrimaryOutput.java (hunk 1 above)
Comment (olenagerasimova) [reply src/main/java/com/artipie/rpm/pkg/PrimaryOutput.java:36]: @g4s8 no, this todo is basically only for a new ctor, that can accept existing not empty tmp file and test for it. I'll implement usage of this new `XmlPrimaryMaid` in a separate PR.

**Comment by olenagerasimova** [comment]
@g4s8 I've made corrections, please, take a look

**Comment by 0crat** [comment]
Job was finished in 3 hours, bonus for fast delivery is possible (see [§36](http://www.zerocracy.com/policy.html#36))

**Pull request merged**